### ISSUE TITLE
[HOLD for payment 2025-02-04] [HOLD for payment 2025-01-24] [$250] mWeb/Safari - BA - The confirmation button jumps when returning to the previous page

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53814]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.73-6
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** n/a
**Email or phone of affected tester (no customers):** n/a
**Issue reported by:** Applause Internal Team

## Action Performed:
Precondition: Navigate to the add bank account modal in Workspace settings
Start and follow the flow for adding a Verifying bank account using the credentials found below until you reach the Company info page

1. Activate the compose box
2. Tap the back arrow button
 
Use Regions Bank
Username: user_good | Password: pass_good
Account type: Plaid Saving (2nd option)
CompanyName: Alberta Bobbeth Charleson
CompanyTaxID: 123456789
First Name: Alberta
Last Name: Charleson

## Expected Result:
The confirmation button doesn't jump when returning to the previous page

## Actual Result:
The confirmation button jumps when returning to the previous page

## Workaround:
Unknown

## Platforms:
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/39f23a26-42a8-492e-a219-2920207c1e9e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021869122959863484757</li>
        <li>Upwork Job ID: 1869122959863484757</li>
        <li>Last Price Increase: 2024-12-24</li>
<li>Automatic offers: </li>
<ul>
<li>huult | Contributor | 105503980</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @CortneyOfstad</details>

**Comment by CortneyOfstad** [comment]
I've been trying to test this and I cannot get it to consistently be recreated, so going to see if this can be retested.

**Comment by CortneyOfstad** [comment]
Still waiting on this being retested!

**Comment by MrMuzyk** [comment]
I am Michał from Callstack - expert contributor group. I’d like to work on this job.

**Comment by CortneyOfstad** [comment]
Hey @mananjadhav! Looks like we have a proposal above for review!

Also, just a heads up that I will be OOO starting this afternoon (December 20th) and will be returning January 6th. A handful of folks on the BZ team will be online for a few days in between the 25th and the 1st, but we'll be operating with a skeleton crew. If any action is needed from a BZ perspective, please post this issue in #expensify-open-source and someone on the team will jump in. Otherwise I will review when I'm back in office. 

Thank you and Happy Holidays!

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The confirmation button jumps when returning to the previous page

### What is the root cause of that problem?

When we tap the back button or outside the input, the keyboard will dismiss, and the `maxHeight` for `KeyboardAvoidingView` will be updated

https://github.com/Expensify/App/blob/a8600a190ae7e83e6299d614d66f560213fb809e/src/components/ScreenWrapper.tsx#L163

https://github.com/Expensify/App/blob/a8600a190ae7e83e6299d614d66f560213fb809e/src/components/ScreenWrapper.tsx#L276-L280

For mWeb Safari on iOS, when the keyboard is dismissed, there is a slight delay in updating the visual viewport. As a result, the `maxHeight` is not updated immediately. Due to this, the button shifts slightly, and after the `maxHeight` is updated, it moves to the bottom, causing a jump as described in this ticket.

### What changes do you think we should make in order to solve the problem?


As I found on [Stack Overflow](https://stackoverflow.com/questions/72747030/delay-on-resize-event-on-virtual-keyboard-on-ios-safari), as of today, there are no real solutions for this, so I suggest a workaround solution to fix the issue.

We should set the `maxHeight` for `KeyboardAvoidingView` when the keyboard is open. If the keyboard is not open, we should not set it, or if the visual viewport is fully updated, then set the `maxHeight` for `KeyboardAvoidingView`. To implement this, follow these steps:

1. Create new state:

```diff
    const [shouldSetMaxHeight, setShouldSetMaxHeight] = useState(true);
```

2. Listen for the input blur or focus events to set setShouldSetMaxHeight. This is an example of how I will find the current way of listening to these events in the source code when making a pull request.

```diff
    useEffect(() => {
        if (!Browser.isSafari()) {
            return;
        }
        const handleFocus = () => {
            setShouldSetMaxHeight(true);
        };

        const handleBlur = () => {
            setShouldSetMaxHeight(false);
        };

        const inputs = document.querySelectorAll('input, textarea');

        inputs.forEach((input) => {
            input.addEventListener('focus', handleFocus);
            input.addEventListener('blur', handleBlur);
        });

        return () => {
            inputs.forEach((input) => {
                input.removeEventListener('focus', handleFocus);
                input.removeEventListener('blur', handleBlur);
            });
        };
    }, []);
```

3. Update the condition to set `maxHeight` for `KeyboardAvoidingView`

https://github.com/Expensify/App/blob/a8600a190ae7e83e6299d614d66f560213fb809e/src/components/ScreenWrapper.tsx#L276-L280

Update to:

```diff
    <KeyboardAvoidingView
        style={[
            styles.w100,
            styles.h100,
            shouldSetMaxHeight ? {maxHeight} : undefined,
            isAvoidingViewportScroll ? [styles.overflowAuto, styles.overscrollBehaviorContain] : {},
        ]}
        behavior={keyboardAvoidingViewBehavior}
        enabled={shouldEnableKeyboardAvoidingView}
    >
```

4. Set `shouldSetMaxHeight` to `true` when the visual viewport is fully updated.

```diff
    useEffect(() => {
        const updateDimensions = () => {
            if (!Browser.isSafari()) {
                return;
            }
            setShouldSetMaxHeight(true);
        };

        const removeViewportResizeListener = addViewportResizeListener(updateDimensions);

        return () => {
            removeViewportResizeListener();
        };
    }, []);
```

<details>
<summary>POC</summary>

https://github.com/user-attachments/assets/bdc6c02f-c33e-4a57-b7ea-94119cdc1b15

</details>

## [Test branch](https://github.com/Expensify/App/compare/main...huult:App:53814-test?expand=1)

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?


This is a UI bug; I don’t think code testing is needed here

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by mananjadhav** [comment]
Reviewing this by tomorrow.

**Comment by mananjadhav** [comment]
I think @huult's RCA is correct and the [proposal](https://github.com/Expensify/App/issues/53814#issuecomment-2559107649) seems to be promising. But I am not sure if we'd want to implement a workaround. Taking feedback from the internal engineer on this one.

🎀 👀 🎀 C+ reviewed.

**Comment by huult** [comment]
Thank you! I will create the pull request soon.

**Comment by CortneyOfstad** [comment]
@mananjadhav can you complete the checklist before EOD? Thanks!

**Comment by mananjadhav** [comment]
@CortneyOfstad The original PR was reverted, and the new PR has been merged on staging. We'll payout based on the new PR.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by yuwenmemon** [comment]
See: https://github.com/Expensify/App/pull/55494#issuecomment-2617217274

We should not do a payout for this solution until @huult can update the fix for Android since it's not working there yet.

**Comment by mananjadhav** [comment]
@huult In case you missed out on the previous comment. I did not come across it in my testing. But I didn't test it on a physical device. 

> See: [#55494 (comment)](https://github.com/Expensify/App/pull/55494#issuecomment-2617217274)
> 
> We should not do a payout for this solution until [@huult](https://github.com/huult) can update the fix for Android since it's not working there yet.

**Comment by mananjadhav** [comment]
@huult Quick bump on the previous comment.

**Comment by huult** [comment]
@mananjadhav Same as you, I am unable to reproduce it on my side

https://github.com/user-attachments/assets/114750d2-cc12-4098-b081-a4ee351679e0

**Comment by CortneyOfstad** [comment]
@huult @mananjadhav @yuwenmemon — is this good for payment now, or is additional action still needed? Thanks for the insight!

**Comment by mananjadhav** [comment]
@yuwenmemon @IuliiaHerets Any updates on the previous comment?

**Comment by IuliiaHerets** [comment]
Hi, @yuwenmemon. Tester still can repro this issue on the mWeb/Safari in the latest build 

https://github.com/user-attachments/assets/157328f6-62b2-481f-9af6-fbabd9896e59

**Comment by huult** [comment]
@IuliiaHerets The video above has fixed the issue. What issue did you mention?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mananjadhav** [comment]
@yuwenmemon Do you still think this should be held?

**Comment by yuwenmemon** [comment]
Doesn't seem like it - @CortneyOfstad this should be good to go!

**Comment by mananjadhav** [comment]
@CortneyOfstad This is ready for payout, can you please post the payment summary. Please note this one had a regression, so the payout would be 125$

**Comment by CortneyOfstad** [comment]
## Payment Summary 

@huult – paid $250 via Upwork
@mananjadhav — to be paid $250 via NewDot

## Regression Test

https://github.com/Expensify/Expensify/issues/469288
